---
title: "Guide to using CAMERA"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{development}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  cache = TRUE
)
```

The following analyses should run within a couple of minutes, depending on internet speed and the traffic that the [IEU GWAS database](https://gwas.mrcieu.ac.uk/) servers are experiencing. Begin by choosing an exposure and outcome hypothesis that can be tested across the populations of interest (e.g. body mass index and coronary heart disease in European ancestry and East Asian ancestry). 


## Data setup

To install CAMERA:

```r
install.packages("devtools")
devtools::install_github("universe77/CAMERA")
```

Open R6 class environment to run CAMERA. Plink (ver.1.90) and LD reference data are also required to identify instruments that can be used for both populations: http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz 


```{r setup}
library(CAMERA)

# Need to specify path to plink and LD reference data here.
# The following exposures and the outcomes will be used: Exposures: BMI (EUR-EAS); Outcomes: CHD (EUR-EAS)
x <- CAMERA$new(
  		exposure_ids=c("ieu-a-2", "bbj-a-1"), 
  		outcome_ids=c("ieu-a-7", "bbj-a-109"), 
  		pops = c("EUR", "EAS"),
  		bfiles=c("/Users/yc16575/OneDrive - University of Bristol/EUR", "/Users/yc16575/OneDrive - University of Bristol/EAS"), 		
  		plink = genetics.binaRies::get_plink_binary(), 		
  		radius=50000, 
  		clump_pop="EAS"
	)
```

Make sure that the exposures/outcomes are matched across the ancestries. Population 1 and 2 should have the same exposure-outcome pair. These data can be extracted from the IEU GWAS database using the [TwoSampleMR](https://mrcieu.github.io/TwoSampleMR/) package. A list of available traits can be obtained using:
```{r}
traits <- TwoSampleMR::available_outcomes()
```

## Running CAMERA

We can now perform the analysis, which will do the following:
1. Extract instruments for the exposures
2. Check the validity of the instruments between the populations (Standardise the data if necessary)
3-1. Extract new instruments from the regions that includes the original instruments from the step 1.
3-2. Extract instruments based on fine-mapping (on development)
4. Extract instruments for the outcomes
5. Harmonise the exposure data and the outcome data
6. Perform MR using genomic structural equation model


## Step by step analysis 

See the `?CAMERA` for options on the parameters for this analysis. 


### Step 1. Extract instruments for the exposures from each pouplation
Note that the exposures/outcomes should be matched across the ancestries. 

```{r}
x$extract_instruments()
```


### Step 2. Check the validity of the instruments
The degree of agreement in instrument associations (instrument-trait) between populations need to be assessed.

```{r}
x$check_phenotypes(ids=x$exposure_ids)
x$check_phenotypes(ids=x$outcome_ids)
```
A value close to 1 indicates that the instrument-trait associations are consistent between the populations. If the values is too low or too high, it is recommended to use other available instruments for the traits (https://gwas.mrcieu.ac.uk/). If the units for the exposures/the outcomes are different across the populations, run:
```{r}
x$standardise_units(exp=x$instrument_raw, out=NULL)
```
You can check the degree of agreement again using:
```{r}
x$check_phenotypes(ids=NULL, after_standardised=x$standardised_exposure)
```
Please note that standardisation is not required if the effect size of SNP-trait is in the unit of SD or log odds.

Then estimate what fraction of the instruments (raw instrument) we would expect to replicate across the populations.
```{r}
x$estimate_instrument_specificity(x$instrument_raw)
x$estimate_instrument_specificity(x$standardised_exposure)
```


### Step 3. Extract new instruments from the regions that includes the original instruments
The following functions will identify region that includes the original instrument. Then it will extract new instruments that are best associated in both populations from the identified regions. Instruments will be selected based on magnitude of instrument-exposure association.

```{r, fig.width = 7, fig.height = 6, out.width = "700px", out.height="600px", dpi=300}
# Identify regions that include the instruments
suppressMessages(x$extract_instrument_regions())

# Extract instrument from identified region by searching across region for best associations
x$scan_regional_instruments()

# Plot regions and their region-selected instruments
x$plot_regional_instruments(instruments=x$instrument_maxz)
```

Check the replication fraction between the populations using the new instruments (instrument_maxz).
```{r}
# Estimate what fraction of the region-selected instruments we'd expect to replicate
x$estimate_instrument_specificity(x$instrument_maxz)
```


### Step 4. Extract instruments for the outcomes
Get outcome data for a particular set of instruments 
```{r}
x$make_outcome_data(exp=x$standardised_exposure, p_exp=1)
```


### Step 5. Harmonise the exposure data and the outcome data
```{r}
x$harmonised_dat(exp=x$standardised_exposure, out=x$instrument_outcome)
```

### Step 6. Perform MR using genomic structural equation model (MR-SEM)
```{r}
x$perform_basic_sem()
```

You can also perform MR-SEM using the instruments from the Step 3.
```{r}
# Get outcome data for a particular set of instruments and run SEM model 
# Standardise the betas and SEs if required
x$standardise_units(exp=x$instrument_maxz)
x$make_outcome_data(exp=x$standardised_exposure, p_exp=1)
x$harmonised_dat(exp=x$standardised_exposure, out=x$instrument_outcome)
x$perform_basic_sem()
```


### Extract SNPs based on fine-mapping
```{r}
suppressWarnings(x$regional_ld_matrices())
suppressWarnings(x$susie_finemap_regions())
x$estimate_instrument_specificity(x$instrument_susie)
x$standardise_units(exp=x$instrument_susie)
x$make_outcome_data(exp=x$standardised_exposure, p_exp=1)
x$harmonised_dat(exp=x$standardised_exposure, out=x$instrument_outcome)
x$perform_basic_sem()
```
